awx/main/migrations/_rbac.py

```diff
@@ -35,7 +38,118 @@ def create_roles(apps, schema_editor):
+    'adhoccommand': 'inventory'
+}
+
+
+def implicit_org_subquery(UnifiedClass, cls, backward=False):
+    """Returns a subquery that returns the so-called organization for objects
+    in the class in question, before migration to the explicit unified org field.
+    In some cases, this can still be applied post-migration.
+    """
+    if cls._meta.model_name not in UNIFIED_ORG_LOOKUPS:
+        return None
+    cls_name = cls._meta.model_name
+    source_field = UNIFIED_ORG_LOOKUPS[cls_name]
+
+    unified_field = UnifiedClass._meta.get_field(cls_name)
+    unified_ptr = unified_field.remote_field.name
+    if backward:
+        qs = UnifiedClass.objects.filter(**{cls_name: OuterRef('id')}).order_by().values_list('tmp_organization')[:1]
+    elif source_field is None:
+        qs = cls.objects.filter(**{unified_ptr: OuterRef('id')}).order_by().values_list('organization')[:1]
+    else:
+        intermediary_field = cls._meta.get_field(source_field)
+        intermediary_model = intermediary_field.related_model
+        intermediary_reverse_rel = intermediary_field.remote_field.name
+        qs = intermediary_model.objects.filter(**{
+            # this filter leverages the fact that the Unified models have same pk as subclasses.
+            # For instance... filters projects used in job template, where that job template
+            # has same id same as UJT from the outer reference (which it does)
+            intermediary_reverse_rel: OuterRef('id')}
+        ).order_by().values_list('organization')[:1]
+    return Subquery(qs)
+
+
+def _migrate_unified_organization(apps, unified_cls_name, backward=False):
+    """Given a unified base model (either UJT or UJ)
+    and a dict org_field_mapping which gives related model to get org from
+    saves organization for those objects to the temporary migration
+    variable tmp_organization on the unified model
+    (optimized method)
+    """
+    start = time()
+    UnifiedClass = apps.get_model('main', unified_cls_name)
+    ContentType = apps.get_model('contenttypes', 'ContentType')
+
+    for cls in UnifiedClass.__subclasses__():
+        cls_name = cls._meta.model_name
+        if backward and UNIFIED_ORG_LOOKUPS.get(cls_name, 'not-found') is not None:
+            logger.debug('Not reverse migrating {}, existing data should remain valid'.format(cls_name))
+            continue
+        logger.debug('Migrating {} to new organization field'.format(cls_name))
+
+        sub_qs = implicit_org_subquery(UnifiedClass, cls, backward=backward)
+        if sub_qs is None:
+            logger.debug('Class {} has no organization migration'.format(cls_name))
+            continue
+
+        this_ct = ContentType.objects.get_for_model(cls)
+        if backward:
+            r = cls.objects.order_by().update(organization=sub_qs)
+        else:
+            r = UnifiedClass.objects.order_by().filter(polymorphic_ctype=this_ct).update(tmp_organization=sub_qs)
```

Comment:
We should verify that this migration is performant (doesn't take forever to finish) and doesn't spike memory if you're performing this operation on, say, tens of millions of `Job` records.

Reply:
@squidboylan let's add this to the plan

Reply:
I fear that performance concerns are much more leaky than the copying of the organization field here. This produces relatively clean SQL code, and we're gotten rid of the OUTER JOINs and such.

The `rebuild_role_parentage` method scales with the number of templates (not the number of jobs), but even so, I think it will be the bottleneck. I am much more hesitant to dive into clever speedups there, because it is using a method which is already called on almost every save the user makes. It is paradoxical that rebuilding _parents_ would take longer than rebuilding _ancestors_, but they do by about 2 orders of magnitude. This is because parents have no database-only representation, and are tied into the model definitions, resulting in frequent trips to the database.

Another performance concern is when you _delete_ an organization. Django's SET_NULL logic has a problem where it does not like to do bulk updates, as we have found with inventory. We have tools to address this if deletion of organizations starts to hit the web request timeout, but like everything else here, it'll require loading up a lot of content.

The same argument applies to job templates, but if organizations are a factor of 5x or 10x worse, that may be enough to cause trouble for the user.